File: packages/python/elizaos/features/basic_capabilities/providers/character.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CHARACTER")


def _resolve_character_placeholders(
    text: str,
    agent_name: str,
    example_names: list[str] | None = None,
) -> str:
    resolved = text.replace("{{agentName}}", agent_name).replace("{{name}}", agent_name)
    for index, name in enumerate(example_names or [], start=1):
        resolved = resolved.replace(f"{{{{name{index}}}}}", name).replace(
            f"{{{{user{index}}}}}", name
        )
    return resolved


def _resolve_list(items: list[str], agent_name: str) -> list[str]:
    return [_resolve_character_placeholders(item, agent_name) for item in items]


async def get_character_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    character = runtime.character
    agent_name = character.name

    sections: list[str] = []

    sections.append(f"# Agent: {agent_name}")

    if character.bio:
        bio_text = character.bio if isinstance(character.bio, str) else "\n".join(character.bio)
        sections.append(f"\n## Bio\n{_resolve_character_placeholders(bio_text, agent_name)}")

    if character.adjectives:
        adjectives = (
            character.adjectives
            if isinstance(character.adjectives, list)
            else [character.adjectives]
        )
        sections.append(
            f"\n## Personality Traits\n{', '.join(_resolve_list(adjectives, agent_name))}"
        )

    # lore is optional and may not exist on all Character instances
    lore = getattr(character, "lore", None)
    if lore:
        lore_text = lore if isinstance(lore, str) else "\n".join(lore)
        sections.append(
            f"\n## Background\n{_resolve_character_placeholders(lore_text, agent_name)}"
        )

    if character.topics:
        topics = character.topics if isinstance(character.topics, list) else [character.topics]
        sections.append(f"\n## Knowledge Areas\n{', '.join(_resolve_list(topics, agent_name))}")

    if character.style:
        style_sections: list[str] = []
        if character.style.all:
            all_style = (
                character.style.all
                if isinstance(character.style.all, list)
                else [character.style.all]
            )
            style_sections.append(f"General: {', '.join(_resolve_list(all_style, agent_name))}")
        if character.style.chat:
            chat_style = (
                character.style.chat
                if isinstance(character.style.chat, list)
                else [character.style.chat]
            )
            style_sections.append(f"Chat: {', '.join(_resolve_list(chat_style, agent_name))}")
        if character.style.post:
            post_style = (
                character.style.post
                if isinstance(character.style.post, list)
                else [character.style.post]
            )
            style_sections.append(f"Posts: {', '.join(_resolve_list(post_style, agent_name))}")
        if style_sections:
            sections.append("\n## Communication Style\n" + "\n".join(style_sections))

    context_text = "\n".join(sections)

    # Note: Protobuf ProviderResult.data is a Struct which has limited type support.
    # The text already contains all the information needed for the agent context.
    return ProviderResult(text=context_text)


character_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_character_context,
    dynamic=_spec.get("dynamic", False),
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/character.rs
//! CHARACTER provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

// Get text content from centralized specs
static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("CHARACTER"));

/// Provider for character information.
pub struct CharacterProvider;

#[async_trait]
impl Provider for CharacterProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(false)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let character = runtime.character();
        let mut sections = Vec::new();

        // Name section
        sections.push(format!("# Agent: {}", character.name));

        // Bio section
        if !character.bio.is_empty() {
            sections.push(format!("\n## Bio\n{}", character.bio));
        }

        // Personality/Adjectives section
        if !character.adjectives.is_empty() {
            sections.push(format!(
                "\n## Personality Traits\n{}",
                character.adjectives.join(", ")
            ));
        }

        // Lore/Background section
        if !character.lore.is_empty() {
            sections.push(format!("\n## Background\n{}", character.lore));
        }

        // Topics/Knowledge areas section
        if !character.topics.is_empty() {
            sections.push(format!(
                "\n## Knowledge Areas\n{}",
                character.topics.join(", ")
            ));
        }

        // Style section
        let mut style_parts = Vec::new();
        if !character.style.all.is_empty() {
            style_parts.push(format!("General: {}", character.style.all.join(", ")));
        }
        if !character.style.chat.is_empty() {
            style_parts.push(format!("Chat: {}", character.style.chat.join(", ")));
        }
        if !character.style.post.is_empty() {
            style_parts.push(format!("Posts: {}", character.style.post.join(", ")));
        }
        if !style_parts.is_empty() {
            sections.push(format!(
                "\n## Communication Style\n{}",
                style_parts.join("\n")
            ));
        }

        let context_text = sections.join("\n");

        Ok(ProviderResult::new(context_text)
            .with_value("agentName", character.name.clone())
            .with_value("hasCharacter", true)
            .with_data("name", character.name.clone())
            .with_data("bio", character.bio.clone()))
    }
}
